src/Queue.h

```diff
@@ -21,14 +21,130 @@
 //	Entries must be either a pointer to the data or nonzero data with
 //	sizeof(data) <= sizeof(void*).
 
-#include "List.h"
+template<typename T>
+class QueueIterator
+	{
+	T* const entries;
+	int offset;
+	int num_entries;
+	T endptr; // let this get set to some random value on purpose. It's only used
+			  // for the operator[] and operator* cases where you pass something
+			  // off the end of the collection, which is undefined behavior anyways.
+public:
+	QueueIterator(T* entries, int offset, int num_entries) :
+		entries(entries), offset(offset), num_entries(num_entries), endptr() {}
+	bool operator==(const QueueIterator& rhs) { return entries == rhs.entries && offset == rhs.offset; }
+	bool operator!=(const QueueIterator& rhs) { return entries != rhs.entries || offset != rhs.offset; }
+	QueueIterator & operator++() { offset++; return *this; }
+	QueueIterator operator++(int) { auto t = *this; offset++; return t; }
+	QueueIterator & operator--() { offset--; return *this; }
+	QueueIterator operator--(int) { auto t = *this; offset--; return t; }
+	std::ptrdiff_t operator-(QueueIterator const& sibling) const { return offset - sibling.offset; }
+	QueueIterator & operator+=(int amount) { offset += amount; return *this; }
+	QueueIterator & operator-=(int amount) { offset -= amount; return *this; }
+	bool operator<(QueueIterator const&sibling) const { return offset < sibling.offset;}
+	bool operator<=(QueueIterator const&sibling) const { return offset <= sibling.offset; }
+	bool operator>(QueueIterator const&sibling) const { return offset > sibling.offset; }
+	bool operator>=(QueueIterator const&sibling) const { return offset >= sibling.offset; }
+	T& operator[](int index)
+		{
+		if (index < num_entries)
+			return entries[index];
+		else
+			return endptr;
+		}
+	T& operator*()
+		{
+		if ( offset < num_entries )
+			return entries[offset];
+		else
+			return endptr;
+		}
+	};
 
-class BaseQueue {
+namespace std {
+	template<typename T>
+	class iterator_traits<QueueIterator<T> >
+	{
+	public:
+		using difference_type = std::ptrdiff_t;
+		using size_type = std::size_t;
+		using value_type = T;
+		using pointer = T;
+		using reference = T&;
+		using iterator_category = std::random_access_iterator_tag;
+	};
+}
+
+
+template<typename T>
+class Queue {
```

Comment:
`using value_type = T;`

Reply:
It's there, just down by where the iterator traits are defined.